src/main/java/com/fasterxml/jackson/databind/deser/std/JsonNodeDeserializer.java

```diff
@@ -755,6 +755,12 @@ protected final JsonNode _fromFloat(JsonParser p, DeserializationContext ctxt,
             return _fromBigDecimal(ctxt, nodeFactory, p.getDecimalValue());
         }
         if (ctxt.isEnabled(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS)) {
+            // [databind#4194] Add an option to fail coercing NaN to BigDecimal
+            // Currently, Jackson 2.x allows such coercion, but Jackson 3.x will not
+            if (p.isNaN() && ctxt.isEnabled(JsonNodeFeature.FAIL_ON_NAN_TO_BIG_DECIMAL_COERCION)) {
```

Comment:
@JooHyukKim p.isNaN is broken - I don't understand why this is merged - we need p.isNan to be fixed first - please can we revert this? https://github.com/FasterXML/jackson-core/pull/1135

Reply:
this code can be made to work if the code is moved to the catch block of the try - the try/catch is there because the p.isNaN does not work reliably

Reply:
> this code can be made to work if the code is moved to the catch block of the try - the try/catch is there because the p.isNaN does not work reliably

Right, makes sense. Thank you for providing a solution @pjfanning 🙏🏼

Reply:
@pjfanning I looked through `ParserBase.isNaN()` and I am not sure why you think it is broken wrt usage here.
Not only are all tests passing (but there may be gaps in coverage), but I think logic makes sense: when a `NaN` value is encountered, value is immediately assigned with `resetAsNaN()`; otherwise parsing is deferred.
And `ParserBase.isNaN()` should not trigger parsing as double.

One thing to specifically note is that since `JsonNodeDeserializer` gets to call accessors first, there should not be cases for `getDoubleValue()` to get called first.

So... how specifically do you think `isNaN()` is broken wrt this specific use case?